apps/openmw/mwworld/worldimp.cpp

```diff
@@ -2743,64 +2752,66 @@ namespace MWWorld
+        osg::Vec3f hitPosition = actor.getRefData().getPosition().asVec3();
 
-        float dist1 = std::numeric_limits<float>::max();
-        float dist2 = std::numeric_limits<float>::max();
+        // for player we can take faced object first
+        MWWorld::Ptr target;
+        if (actor == MWMechanics::getPlayer())
+            target = getFacedObject();
 
-        if (result1.mHit)
-            dist1 = (origin - result1.mHitPos).length();
-        if (result2.mHit)
-            dist2 = (origin - result2.mHitPointWorld).length();
+        // if the faced object can not be activated, do not use it
+        if (!target.isEmpty() && !target.getClass().canBeActivated(target))
+            target = NULL;
 
-        if (result1.mHit)
-        {
-            target = result1.mHitObject;
-            hitPosition = result1.mHitPos;
-            if (dist1 > getMaxActivationDistance() && !target.isEmpty() && (target.getClass().isActor() || !target.getClass().canBeActivated(target)))
-                target = NULL;
-        }
-        else if (result2.mHit)
+        if (target.isEmpty())
         {
-            target = result2.mHitObject;
-            hitPosition = result2.mHitPointWorld;
-            if (dist2 > getMaxActivationDistance() && !target.isEmpty() && !target.getClass().canBeActivated(target))
-                target = NULL;
-        }
+            // For actor targets, we want to use hit contact with bounding boxes.
+            // This is to give a slight tolerance for errors, especially with creatures like the Skeleton that would be very hard to aim at otherwise.
+            // For object targets, we want the detailed shapes (rendering raycast).
+            // If we used the bounding boxes for static objects, then we would not be able to target e.g. objects lying on a shelf.
+            std::pair<MWWorld::Ptr,osg::Vec3f> result1 = getHitContact(actor, fCombatDistance, targetActors);
 
-        // When targeting an actor that is in combat with an "on touch" spell, 
-        // compare against the minimum of activation distance and combat distance.
+            // Get the target to use for "on touch" effects, using the facing direction from Head node
+            osg::Vec3f origin = getActorHeadTransform(actor).getTrans();
 
-        if (!target.isEmpty() && target.getClass().isActor() && target.getClass().getCreatureStats (target).getAiSequence().isInCombat()) 
-        {
-            distance = std::min (distance, getStore().get<ESM::GameSetting>().find("fCombatDistance")->getFloat());
-            if (distance < dist1)
-                target = NULL;
+            osg::Quat orient = osg::Quat(actor.getRefData().getPosition().rot[0], osg::Vec3f(-1,0,0))
+                    * osg::Quat(actor.getRefData().getPosition().rot[2], osg::Vec3f(0,0,-1));
+
+            osg::Vec3f direction = orient * osg::Vec3f(0,1,0);
+            float distance = getMaxActivationDistance();
```

Comment:
Telekinesis no longer gives a bonus (the distance was previously getActivationDistancePlusTelekinesis()). Was that intentional? Shouldn't on-touch spells benefit from telekinesis?

Reply:
Actually, gives a bonus, when we use the focused object, as in vanilla game.
From my testing, telekinesis does not give distance bonus against other actors (use hitboxes here).

Reply:
Alright. Otherwise everything looks and works fine for me.

Reply:
Though the original change was very much intentional, see [this Allofich's PR](https://github.com/OpenMW/openmw/pull/1074).

Reply:
I do not understand what "explosions for touch spell" means, but for normal touch spells telekinesis still give you range bonus for actors out of combat, because we use the faced object as a target, if you can activate it (tooltip is shown).

Reply:
Oh, I misparsed the code again. Carry on.

Reply:
> I do not understand what "explosions for touch spells" means,

"Explosion" means the effect that happens when a spell has an `Area` value specified and the spell collides with something, like when a fireball hits something. I found that with a touch-range spell, if the spell has a non-0 Area value and is used on some objects (for example, clothes or weapons, or doors that do not transport the player to a new cell), there will also be such an explosion, and the explosion can damage actors. If a Telekinesis effect is active, although you cannot use a touch-range spell on an actor directly, you can use the explosion effect to damage them from range this way.

 > for normal touch spells telekinesis still give you range bonus for actors out of combat
In the original game telekinesis does not give you a range bonus for actors out of combat. Probably so you can't "talk" to them from far away. Because of this, you also can't use a touch-range spell + telekinesis combination on them directly.

Reply:
> In the original game telekinesis does not give you a range bonus for actors out of combat. Probably so you can't "talk" to them from far away

Nevermind, ``getFacedObject()`` does not take in account telekinesis anyway if ``facedObject.getClass().allowTelekinesis(facedObject)`` returns false.
So this PR should replicate described behaviour for now:

> we use the faced object as a target, if you can activate it (tooltip is shown).